ncm-dpmlfc/src/main/perl/dpmlfc/RuleBasedEditor.pm

```diff
@@ -0,0 +1,817 @@
+
+local(*DTA);
+
+# Constants use to format lines in configuration files
+# Exported constants
+use enum qw(LINE_FORMAT_PARAM=1
+            LINE_FORMAT_ENVVAR
+            LINE_FORMAT_XRDCFG
+            LINE_FORMAT_XRDCFG_SETENV
+            LINE_FORMAT_XRDCFG_SET
+           );
+use enum qw(LINE_VALUE_AS_IS
+            LINE_VALUE_BOOLEAN
+            LINE_VALUE_HOST_LIST
+            LINE_VALUE_INSTANCE_PARAMS
+            LINE_VALUE_ARRAY
+            LINE_VALUE_HASH_KEYS
+            LINE_VALUE_STRING_HASH
+           );
+use enum qw(LINE_VALUE_OPT_NONE
+            LINE_VALUE_OPT_SINGLE
+           );
+# Internal constants
+Readonly my $LINE_FORMAT_DEFAULT => LINE_FORMAT_PARAM;
+Readonly my $LINE_QUATTOR_COMMENT => "\t\t# Line generated by Quattor";
+Readonly my $LINE_OPT_DEF_REMOVE_IF_UNDEF => 0;
+Readonly my $LINE_OPT_DEF_ALWAYS_RULES_ONLY => 0;
+Readonly my $RULE_CONDITION_ALWAYS => 'ALWAYS';
+Readonly my $RULE_OPTION_SET_GLOBAL => 'GLOBAL';
+
+
+# Export constants used to build rules
+Readonly my @RULE_CONSTANTS => ('LINE_FORMAT_PARAM',
+                                'LINE_FORMAT_ENVVAR',
+                                'LINE_FORMAT_XRDCFG',
+                                'LINE_FORMAT_XRDCFG_SETENV',
+                                'LINE_FORMAT_XRDCFG_SET',
+                                'LINE_VALUE_AS_IS',
+                                'LINE_VALUE_BOOLEAN',
+                                'LINE_VALUE_HOST_LIST',
+                                'LINE_VALUE_INSTANCE_PARAMS',
+                                'LINE_VALUE_ARRAY',
+                                'LINE_VALUE_HASH_KEYS',
+                                'LINE_VALUE_STRING_HASH',
+                                'LINE_VALUE_OPT_NONE',
+                                'LINE_VALUE_OPT_SINGLE',
+                               );
+our @EXPORT_OK;
+our %EXPORT_TAGS;
+push @EXPORT_OK, @RULE_CONSTANTS;
+$EXPORT_TAGS{rule_constants} = \@RULE_CONSTANTS;
+
+
+# Backup file extension
+use constant BACKUP_FILE_EXT => ".old";
+
+
+=pod
+
+=head1 DESCRIPTION
+
+This module implements a rule-based editor. It has only one public method: B<updateConfigfile>.
+Rules are passed as a hash.
+
+See https://github.com/quattor/CAF/issues/123#issue-123702165 for details.
+
+
+=head2 Public methods
+
+=over
+
+=item updateConfigfile
+
+Update configuration file content,  applying configuration rules.
+
+Arguments :
+    file_name: name of the file to update
+    config_rules: config rules corresponding to the file to build
+    config_options: configuration parameters used to build actual configuration
+    options: a hash setting options to modify the behaviour of this function
+
+Supported entries for options hash:
+    always_rules_only: if true, apply only rules with ALWAYS condition (D: false)
+    remove_if_undef: if true, remove maatching configuration line is rule condition is not met (D: false)
+
+=cut
+
+sub updateConfigFile {
+  my $function_name = "updateConfigFile";
```

Comment:
Same as before, I'll change for 4 space ident later, when everything is clean...